paimon-core/src/main/java/org/apache/paimon/operation/LocalOrphanFilesClean.java

```diff
@@ -113,31 +116,52 @@ public List<Path> clean() throws IOException, ExecutionException, InterruptedExc
         return deleteFiles;
     }
 
-    private List<String> getUsedFiles(String branch) {
-        List<String> usedFiles = new ArrayList<>();
+    private void collectWithoutDataFile(
+            String branch, Consumer<String> usedFileConsumer, Consumer<String> manifestConsumer)
+            throws IOException {
+        randomlyOnlyExecute(
+                executor,
+                snapshot -> {
+                    try {
+                        collectWithoutDataFile(
+                                branch, snapshot, usedFileConsumer, manifestConsumer);
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                },
+                safelyGetAllSnapshots(branch));
+    }
+
+    private Set<String> getUsedFiles(String branch) {
+        Set<String> usedFiles = ConcurrentHashMap.newKeySet();
         ManifestFile manifestFile =
                 table.switchToBranch(branch).store().manifestFileFactory().create();
         try {
-            Set<String> manifests = new HashSet<>();
+            Set<String> manifests = ConcurrentHashMap.newKeySet();
             collectWithoutDataFile(branch, usedFiles::add, manifests::add);
-            List<String> dataFiles = new ArrayList<>();
-            for (String manifestName : manifests) {
-                retryReadingFiles(
-                                () -> manifestFile.readWithIOException(manifestName),
-                                Collections.<ManifestEntry>emptyList())
-                        .stream()
-                        .map(ManifestEntry::file)
-                        .forEach(
-                                f -> {
-                                    if (candidateDeletes.contains(f.fileName())) {
-                                        dataFiles.add(f.fileName());
-                                    }
-                                    f.extraFiles().stream()
-                                            .filter(candidateDeletes::contains)
-                                            .forEach(dataFiles::add);
-                                });
-            }
-            usedFiles.addAll(dataFiles);
+            randomlyOnlyExecute(
```

Comment:
I think there is indeed a risk of OOM.

But this can be avoided by reducing the parallelism (if it's equal to 1, then it's similar to serial execution before), and here it's just providing an ability to read in parallel.

If parallelism 1 is still OOM, then distributed mode is considered.

Reply:
What about change randomlyOnlyExecute to sequentialBatchedExecute?

Reply:
Yeah, I repleaced it with sequentialBatchedExecute since It has better control over memory

Reply:
In the end, you still need to put it in a set, so as long as there is no inflation during the reading process, there should be no risk of oom here.

Reply:
I test ```sequentialBatchedExecute``` ```randomlyOnlyExecute``` and ```randomlyExecute``` on table with 2PB in 16 parallelism, None of them happened OOM. Besides ```sequentialBatchedExecute``` procedure cost **30 min** , ```randomlyOnlyExecute``` and ```randomlyExecute``` cost about **20 min**

**So I prefer to use randomlyOnlyExecute** because this method is more concise and no other variables are introduced, it can also reduce the gc process of temporary data files variable. The disadvantage of this method is that there will be waiting lock overhead while inserting useFiles, but this has no impact on the overall time consumption, because the overall time consumption mainly comes from IO.

What do you think? @JingsongLi @wwj6591812